src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -140,6 +139,29 @@ void Service::processQuery(const HTMLForm & params, ReadBuffer & /*body*/, Write
     {
         part = findPart(part_name);
 
+        /// Ephemeral zero-copy lock may be lost for PreActive parts
+        /// do not expose PreActive parts
+        if (client_protocol_version >= REPLICATION_PROTOCOL_VERSION_WITH_WAITING_PREACTIVE)
```

Comment:
Philosophical question: Why prefer waiting on the "client" (when error is returned) instead of waiting in the server? If we waited in the server we wouldn't need to add a new protocol version, just wait for some time and return it (or an error if waited for more than x amount of time).

Reply:
Possible too.
I just not see it good if server just hungs. Some receive timeout on client could be triggered. That would be looked like either network errors or server hungs. Sending some headers time to time could solve that, but it overkill.
But I agree that this is overthinking.
Will try your approach.